File: javascript/protein-translation/protein-translation.js
const STOP_CODONS = ['UAA', 'UAG', 'UGA'];
const CODONS_BY_PROTEIN = {
    'Methionine': ['AUG'],
    'Phenylalanine': ['UUU', 'UUC'],
    'Leucine': ['UUA', 'UUG'],
    'Serine': ['UCU', 'UCC', 'UCA', 'UCG'],
    'Tyrosine': ['UAU', 'UAC'],
    'Cysteine': ['UGU', 'UGC'],
	'Tryptophan': ['UGG']
};
const PROTEINS = Object.keys(CODONS_BY_PROTEIN);

const splitSequence = (sequence) => {
    const codon = sequence.substring(0, 3);
    const tail = sequence.substring(3);
    return [codon, tail];
}

const getProteins = (codon) => {
    return PROTEINS.find((protein) => CODONS_BY_PROTEIN[protein].includes(codon))
};

const translate = (sequence) => {
    sequence = sequence || '';
    let output = [];
    let protein;
    while (sequence.length) {
        [codon, sequence] = splitSequence(sequence);
        if (STOP_CODONS.includes(codon)) {
            break;
        }
        protein = getProteins(codon);
        if (protein) {
            output.push(protein);
        } else {
            throw new Error('Invalid codon');
        }
    }
    return output;
};

module.exports = translate; 

File: fsharp/protein-translation/ProteinTranslation.fs
module ProteinTranslation

open System

let private STOP_CODONS = ["UAA"; "UAG"; "UGA"];
let private CODONS_BY_PROTEIN = 
    Map.ofList [
        ("Methionine", ["AUG"]);
        ("Phenylalanine", ["UUU"; "UUC"]);
        ("Leucine", ["UUA"; "UUG"]);
        ("Serine", ["UCU"; "UCC"; "UCA"; "UCG"]);
        ("Tyrosine", ["UAU"; "UAC"]);
        ("Cysteine", ["UGU"; "UGC"]);
        ("Tryptophan", ["UGG"]) ]

let getProtein (codon: string): string =
    Map.fold (fun state protein codons -> if List.contains codon codons then protein else state) "" CODONS_BY_PROTEIN

let slices (str: string) (length: int): string list = 
    if str.Length < 1 || length < 1 || length > str.Length
    then []
    else
        [0 .. length .. str.Length - length]
        |> List.map (fun i -> str.[i .. i + length - 1])

let proteins (rna: string): string list = 
    slices rna 3
    |> Seq.takeWhile (fun c -> not <| List.contains c STOP_CODONS) 
    |> List.ofSeq
    |> List.map getProtein